tests/test_cufgen.py

```diff
@@ -0,0 +1,117 @@
+import pytest
+
+from conftest import available_frontends
+from loki import Module, FindNodes, Transformer
+from loki import ir
+from loki.expression import symbols as sym
+
+
+@pytest.mark.parametrize('frontend', available_frontends())
+def test_cufgen(frontend):
+    """
+    A simple test routine to test the Cuda Fortran (CUF) backend
+    """
+
+    fcode = """
+module transformation_module_cufgen
+  implicit none
+  integer, parameter :: len = 10
+contains
+
+  subroutine driver(a, b, c)
+    integer, intent(inout) :: a
+    integer, intent(inout) :: b(len)
+    integer, intent(inout) :: c(a, len)
+    integer :: var_device
+    integer :: var_managed
+    integer :: var_constant
+    integer :: var_shared
+    integer :: var_pinned
+    integer :: var_texture
+    call kernel(a, b)    
+  end subroutine driver
+
+  subroutine kernel(a, b)
+    integer, intent(inout) :: a
+    integer, intent(inout) :: b(len)
+    real :: x(a) 
+    real :: k2_tmp(a, a)
+    call device1(x, k2_tmp) 
+  end subroutine kernel
+
+  subroutine device(x, y)
+    real, intent(inout) :: x(len)
+    real, intent(inout) :: y(len, len)
+  end subroutine device
+
+end module transformation_module_cufgen
+"""
+
+    module = Module.from_source(fcode, frontend=frontend)
+
+    driver = None
+    kernels = []
+    device_subroutines = []
+    for routine in module.routines:
+        if "driver" in routine.name:
+            driver = routine
+        elif "kernel" in routine.name:
+            kernels.append(routine)
+        elif "device" in routine.name:
+            device_subroutines.append(routine)
+
+    assert driver
+    assert module.to_fortran(cuf=True) == module.to_fortran()
+
+    decl_map = {}
+    for decl in FindNodes(ir.VariableDeclaration).visit(driver.spec):
```

Comment:
To make you aware of a simpler way of doing this: You can simply edit the variable types directly:
```
for var in driver.variables:
    if "device" in var.name:
        var.type = var.type.clone(device=True)
    if "managed" in var.name:
        ...
```

Reply:
Yes I tried this ... 
However, I don't know why, but `var.type = var.type.clone(device=True)` does **not** work *AttributeError: can't set attribute 'type'*.

```py
for var in driver.variables:
    if "device" in var.name:
        var.clone(type=var.type.clone(device=True))
```

works, but probably shouldn't work.

Reply:
Then there's probably a `type` setter missing on one of the variable classes. What type does `var` have in that context.
The second works but is unnecessary - it uses the same mechanism of updating the symbol table under the hood.

Reply:
Yes, the `MetaSymbol` class is lacking this/those setter(s) (in general).

Adding

```py
    @type.setter
    def type(self, _type):
        if self.symbol.scope is None:
            # Store locally if not attached to a scope
            self.symbol.type = _type
        elif _type is None:
            # Store deferred type if unknown
            self.symbol.scope.symbol_attrs[self.symbol.name] = SymbolAttributes(BasicType.DEFERRED)
        elif _type is not self.symbol.scope.symbol_attrs.lookup(self.symbol.name):
            # Update type if it differs from stored type
            self.symbol.scope.symbol_attrs[self.symbol.name] = _type
```

to the `MetaSymbol` class solves the issue. However, all other properties in `MetaSymbol` do not have setters as well.

Reply:
Yes, IR and expression nodes have intentionally been kept immutable. We may consider relaxing this gradually but it easily risks ending up in an inconsistent state.
However, `type` property is somewhat separate to that because it modifies the symbol table of the enclosing scope rather than the node itself, which is why this has been added recently. `MetaSymbol` seems to have been forgotten there.
However, there's no need to re-implement the logic, simply use the setter on `self.symbol`:
```python
@type.setter
def type(self, _type):
    self.symbol.type = _type
```